### ISSUE TITLE
ABC size: 'B' metric harshness

### ISSUE THREAD

**Issue by Kache** [bbatsov/rubocop#3335]
Discovered that the ABC size is very harsh when counting 'B' (method calls).

Since we're using Ruby, the following are all "actually" method calls:

```ruby
obj_instance.accessor
hash_or_array[value_key]
```

Methods like `foo` below have a surprisingly high ABC score (16.12):

```ruby
class Foo # :nodoc:
  def foo
    hash[bar] += aux(bar.nil?, baz.nil?)
    hash[baz] += aux(baz.nil?, bar.nil?)
  end

  private

  attr_reader :hash, :bar, :baz

  def aux(a, b)
  end
end
```

Is there a way to make it so the following are either considered zero or half cost? (e.g. local variables are cost 1 to assign, but zero to access):

 * an instance variable "passthrough" accessor method, e.g. accessors created via `attr_reader` or ones that have a syntax like:
```ruby
def plain_accessor
  @plain_accessor ||= 'value'.freeze
end
```
 * `[]` access for whitelisted abstract data container types (Hash, Array, Set, configurable others?)

Because of the "B" harshness, the current ABC metric subtly encourages things like trading "B" counts for expressions that fall under "A" or "C" and using instance variables directly (avoiding OO interfacing), e.g. (also contrived):

```ruby
# ABC of 8.60, but via "score manipulation" and also far less desirable, IMO
def bar
  local_baz = baz
  hash[@bar] += aux(@bar || nil, local_baz || nil)
  hash[local_baz] += aux(local_baz || nil, @bar || nil)
end
```

**Comment by jonas054** [comment]
I kind of agree with your point about accessor methods. If we only made the change that calling methods defined by `attr_...` wouldn't count, then the ABC score for `foo` would be the same as for
```ruby
  def foo
    hash[@bar] += aux(@bar.nil?, @baz.nil?)
    hash[@baz] += aux(@baz.nil?, @bar.nil?)
  end
```
which is 10.2. There is a problem, though, and that's inheritance and mixins. We would only be able to detect `attr_...` calls in the same file. So I wonder if it's worth it.

Your other points, the ones about calling simple methods with a certain syntax, and making exceptions for certain calls to `[]` and `[]=`, I don't agree with. Those are pretty normal method calls that should be counted IMO.

All in all, I come down in favor of keeping things as they are.

**Comment by bbatsov** [comment]
Agreed.

**Comment by Kache** [comment]
Rubocop is particularly being annoying in this regard in my tests since every `assert` and `refute` (and the other methods starting with the same string, `assert_equal`, etc) adds to the already harsh B-count.

It'd be nice to be able to configure not counting asserts while in the test directory.

**Comment by Drenmi** [comment]
@Kache: The decision is to not account for any tools or frameworks in RuboCop core, and defer that to extensions once an extension system is in place, so for now you'd need to disable the cop for your test files.

**Comment by jonas054** [comment]
> I can just completely disable the ABC size cop for tests as a workaround

That's a very old instruction for how to disable something locally that's you're quoting. Now you can just add the following to your `.rubocop.yml`.

```yaml
Metrics/AbcSize:
  Exclude:
    - 'test/**/*'
```

> If while, until, and for are supposed to capture loops, what about counting the common Enumerable iteration methods, each, map, select, etc as C instead?

I prefer to count `each` as a method call, because that's what it is. Trying to reassign some method calls to other categories is adding needless complexity, and would probably generate a long string of new issues down the road.

There's is, however, one aspect of how we count the size of a `for` compared to an `each` that needs some improvement. Take these methods:

```ruby
def use_each
  @collection.each do |a, b|
  end
end

def use_for
  for a, b in @collection
  end
end
```

They are parsed as
```lisp
(def :use_each
  (args)
  (block
    (send
      (ivar :@collection) :each)
    (args
      (arg :a)
      (arg :b)) nil))

(def :use_for
  (args)
  (for
    (mlhs
      (lvasgn :a)
      (lvasgn :b))
    (ivar :@collection) nil))
```

and that gives us
```
A=0, B=1, C=0 => 1.0 (each)
A=2, B=0, C=1 => 2.24 (for)
```

It seems to me that we should either count block parameters as assignments, or remove the assignment nodes in the `for` statement from the ABC size.
